### DISCUSSION TITLE
openvdb::tools::foreach iterating through all the voxels in each leaf

### DISCUSSION THREAD

**Discussion by drperpen** [AcademySoftwareFoundation/openvdb#1540] [Q&A] [answered]
Hi,
I am trying to do a basic volume carving with matte/silhouette images and known camera poses/intrinsics. Everything works fine if I use an accessor and check every single value inside a triple loop. Then I aimed to speed things up using the multithreaded openvdb::tools::foreach minimal example in the docs:

```
struct Local {
    static inline void op(const FloatGrid::ValueAllIter& iter) {
        iter.setValue(*iter * 2);
    }
};
FloatGrid grid = ...;
tools::foreach(grid.beginValueAll(), Local::op);
```
The result I get is not a detailed as the first option, it looks like it's not looping though all the voxels, but just the leafs, and assigns the value of the leaf to all its voxels. How could iterate through the voxels of each Leaf iteration?
Thanks!

**Comment by Idclip** [comment, marked as answer]
Hi @drperpen,

When you say "a triple loop" I assume you're looping over a rectangular domain in the form:

```c++
for (int x = minx; x < maxx; ++x)
  for (int y = miny; y < maxy; ++y)
    for (int z = minz; z < maxz; ++z)
      // acc.setValue() ?
```

If you're then using accessor methods like `setValue()` then this may produce different behaviour to your `::foreach` example. Importantly the `::foreach` solution will not modify the volumes topology where as the loop solution will be a) implicitly creating new nodes if they didn't previously exist at the location `(x,y,z)` and b) densifying tiles to leaf nodes which, from the looks of things, is unnecessary in this case. I'm not sure how you're visualizing the output but the "less detail" you're seeing is most likely a combination of these two things.

If you actually need a fully dense VDB in the min->max coordinate domain then, in general, you must make sure this topology exists before iterating on it when using multithreaded methods. For example:

```c++
// Create a mask topology of the domain we need to exist
CoordBBox box(Coord(minx,miny,minz), Coord(maxx,maxy,maxz))
MaskGrid mask;
mask.sparseFill(box, /*value=*/true);

FloatGrid grid = ...;
// Union our grids topology with the mask domain. This makes sure that all node topologies exist but preserves our grids values
grid.topologyUnion(*mask);
mask.clear();

// Forcefully densify the entire VDB (only necessary if you know you need leaf level values)
/* grid.tree().voxelizeActiveTiles() */

// run op
tools::foreach(grid.beginValueAll(), Local::op);
```

**Reply by drperpen** [reply]
Thanks for your answer, it definitely pointed in the right direction. I was filling my grid with:
`grid->fill(bbox, 1.0, 1);`
After reading your message I changed it to:
`grid->denseFill(bbox, 1.0, 1);`
And it works as I expented. When I check the result in Houdini, I am getting "voxel" detail level with my carve, and not "leaf" detail level as before. I appreciate your help!
Daniel
